src/Compilers/Core/Portable/StrongName/CryptoBlobParser.cs

```diff
@@ -181,62 +182,136 @@ private static ImmutableArray<byte> CreateSnPublicKeyBlob(byte type, byte versio
         /// <remarks>
         /// Can be either a PUBLICKEYBLOB or PRIVATEKEYBLOB. The BLOB must be unencrypted.
         /// </remarks>
-        public unsafe static bool TryGetPublicKey(ImmutableArray<byte> blob, out ImmutableArray<byte> publicKey)
+        public static bool TryParseKey(ImmutableArray<byte> blob, out ImmutableArray<byte> snKey, out RSAParameters? privateKey)
         {
-            // Is this already a strong name PublicKeyBlob?
+            privateKey = null;
+            snKey = default(ImmutableArray<byte>);
+
+            var asArray = blob.DangerousGetUnderlyingArray();
+
             if (IsValidPublicKey(blob))
             {
-                publicKey = blob;
+                snKey = blob;
                 return true;
             }
 
-            publicKey = ImmutableArray<byte>.Empty;
-
-            // Must be at least as large as header + RSA info
             if (blob.Length < BlobHeaderSize + RsaPubKeySize)
             {
                 return false;
             }
 
-            fixed (byte* ptr = blob.DangerousGetUnderlyingArray())
+            try
             {
-                var blobReader = new BlobReader(ptr, blob.Length);
-
-                // Header (corresponds to PUBLICKEYSTRUC struct in wincrypt.h)
-                var type = blobReader.ReadByte();
-                var version = blobReader.ReadByte();
-                var reserved = blobReader.ReadUInt16();
-                var algId = blobReader.ReadUInt32();
-
-                // Info (corresponds to RSAPUBKEY struct in wincrypt.h)
-                var magic = blobReader.ReadUInt32();
-                var bitLen = blobReader.ReadUInt32();
-                var pubExp = blobReader.ReadUInt32();
-
-                var modulusLength = (int)(bitLen >> 3);
-                // The key blob data just contains the modulus
-                if (blob.Length - s_offsetToKeyData < modulusLength)
-                {
-                    return false;
+                using (MemoryStream memStream = new MemoryStream(asArray))
```

Comment:
Can we abstract out the copied code to a separate function, even local function would be fine, and add a comment on where we got it from.